### PR TITLE
Add installcheck phase to AutotoolsPackage

### PR DESCRIPTION
**Pull request by adamjstewart**
From the [automake documentation](https://www.gnu.org/software/automake/manual/automake.html):

> A last and optional step is to run make installcheck. This command may run tests on the installed files. make check tests the files in the source tree, while make installcheck tests their installed copies. The tests run by the latter can be different from those run by the former. For instance, there are tests that cannot be run in the source tree. Conversely, some packages are set up so that make installcheck will run the very same tests as make check, only on different files (non-installed vs. installed). It can make a difference, for instance when the source tree’s layout is different from that of the installation. Furthermore it may help to diagnose an incomplete installation.
>
> Presently most packages do not have any installcheck tests because the existence of installcheck is little known, and its usefulness is neglected.

It isn't very common, but it is a real phase.

### CODE REVIEW

**Comment by adamjstewart** [comment]
@alalazo I've been thinking about this while looking at `AutotoolsPackage` and my own `PythonPackage`. Right now, we have something that looks like this:
```python
phases = ['autoreconf', 'configure', 'build', 'install']

def autoreconf
def configure
def build
def install

@PackageBase.sanity_check('build')                                        
@PackageBase.on_package_attributes(run_tests=True)                          
def check():

@PackageBase.sanity_check('install')                                        
@PackageBase.on_package_attributes(run_tests=True)                          
def installcheck
```
How would you feel about changing this to:
```python
phases = ['autoreconf', 'configure', 'build', 'install']
if self.run_tests:
    phases = ['autoreconf', 'configure', 'build', 'check', 'install', 'installcheck']

def autoreconf
def configure
def build
def check
def install
def installcheck
```

**Comment by alalazo** [comment]
> How would you feel about changing this ...

@adamjstewart I am not very fond of the idea for `AutotoolsPackage`. What I had in mind when I implemented #1186 was that phases should have been fixed for a given base class and **always** executed by spack during installation. That's why I excluded `check`. To be honest I was even unsure whether or not to add `autoreconf`. 

The hooks that are available through the decorators permit you to execute code either before or after a phase, but the conceptual difference for me is that a phase is always executed while `check` or `installcheck` can be skipped.

I understand though that for `PythonPackage` the situation is different.

-----

Minor point: we don't have `self` at class scope during class definition. But we can work around this implementation issue.

**Comment by adamjstewart** [comment]
I understand the semantic difference. I guess the main user-facing difference is that if something crashes during `check`, Spack will tell the user that it crashed during `build`. I think it would be nice for the user to know whether it actually crashed during build or check.

**Comment by alalazo** [comment]
@adamjstewart I get the point, and I think we have an issue open to modify the `tty` logger to let selected messages pass during the execution of a phase (as currently they are redirected if not in `verbose` mode).

**Comment by adamjstewart** [comment]
Ping @tgamblin

File: lib/spack/spack/build_systems/autotools.py [hunk 1]
```diff
@@ -288,5 +291,14 @@ def check(self):
             self._if_make_target_execute('test')
             self._if_make_target_execute('check')
 
+    run_after('install')(PackageBase._run_default_install_time_test_callbacks)
+
+    def installcheck(self):
+        """Searches the Makefile for an ``installcheck`` target
+        and runs it if found.
+        """
+        with working_dir(self.build_directory()):
```
Comment (alalazo) [inline lib/spack/spack/build_systems/autotools.py]: @adamjstewart I think it should read `self.build_directory` as it is a property now.

**Comment by tgamblin** [comment]
@alalazo @adamjstewart: I want to merge this, but if the `installcheck` name changes, are people going to be torn up about that?

@BarrySmith and I are talking about how we can better and more consistently support testing in Spack, specifically:
1. Smoke tests
1. Longer tests that one might not want to run immediately after install

Ideally we'd have a spack command for each of those (or a couple subcommands, or options -- something).  And ideally, we could run those commands at an arbitrary time after an install.  However, depending on the project, there can be two-ish kinds of tests:

1. Tests that run out of the build directory
1. Tests that run on an installation directory

Obvious issue: the build directory gets removed after a successful install, so (1) is difficult. I'm thinking about ways to make that customizable for the user, or maybe to have a lifetime imposed on build stages.

Anyway, back on topic, I suspect there will eventually be standard commands for running smoke tests and long tests.  I would like `installcheck` to eventually become one of those.  I doubt this is going to disturb much because it's so infrequently used, but I just wanted to say it before merging.

Any objections?

**Comment by alalazo** [comment]
> Anyway, back on topic, I suspect there will eventually be standard commands for running smoke tests and long tests. I would like installcheck to eventually become one of those. I doubt this is going to disturb much because it's so infrequently used, but I just wanted to say it before merging.
>
> Any objections?

Not from me. 

In this context `installcheck` is a name looked up in the current package. As long as the new command `installcheck` doesn't require to code a new method in the package with the same name the two things are completely decoupled, right?

**Comment by davydden** [comment]
> Anyway, back on topic, I suspect there will eventually be standard commands for running smoke tests and long tests. I would like installcheck to eventually become one of those

i think it's ok to run smoke tests based on `--run-tests` argument to `install`, so they are off by default but can be invoked if requested. For long-time tests (also of interest to `dealii` package), `installcheck` would probably be a good candidate. 

Just so that others are aware, we've been discussing those issues in https://github.com/LLNL/spack/issues/2060

**Comment by davydden** [comment]
@alalazo 

> the new command installcheck doesn't require to code a new method in the package with the same name 

my understanding is that it will most likely require some per-package actions. To give an example of `dealii` (which i know), you would need `cmake`, checkout the `github` repo of `dealii` which has tests and then run
```
cmake -DDEAL_II_DIR=SPACK_ROOT/opt/spack/bla/bla/dealii-develop-oj72i6wetyp7hemjiim3ltdv4kxwkty6 /path/to/dealii_source/tests
```
where `/path/to/dealii_source/tests` is part of repository.

**Pull request merged**